File: app/src/main/java/org/kbieron/iomerge/services/InputDevice.java
package org.kbieron.iomerge.services;

import android.content.Context;
import android.util.Log;

import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;


@EBean(scope = EBean.Scope.Singleton)
public class InputDevice {

	private static final String DAEMON_NAME = "iomerge_daemon";

	static {
		System.loadLibrary("native");
	}

	@RootContext
	protected Context context;

	synchronized protected void startNativeDaemon() throws IOException, InterruptedException {
		stop();

		String outPath = context.getCacheDir().getAbsolutePath() + File.separator + DAEMON_NAME;
		try (FileOutputStream output = new FileOutputStream(new File(outPath))) {
			copy(context.getAssets().open(DAEMON_NAME), output);
		}

		Runtime.getRuntime().exec(new String[]{"su", "-C", "chmod 777 " + outPath}).waitFor();
		initializePipe();

		Runtime.getRuntime().exec(new String[]{"su", "-C", outPath});
	}

	synchronized public void stop() {

		try {
			//noinspection SpellCheckingInspection
			Runtime.getRuntime().exec(new String[]{"su", "-C", "killall " + DAEMON_NAME}).waitFor();
		} catch (InterruptedException | IOException e) {
			e.printStackTrace();
		}
	}

	private native void initializePipe();

	native void mouseMove(int x, int y);

	native void mousePress();

	native void mouseRelease();

	native void keyPress(int c);

	native void keyRelease(int c);

	native void mouseWheel(int anInt);

	void emitKeyEvent(int event) {
		//noinspection SpellCheckingInspection
		String[] execParams = {"su", " -C", "input keyevent " + event};

		try {
			Runtime.getRuntime().exec(execParams);
		} catch (IOException e) {
			Log.e("IOManage", "Unable to run ", e);
		}
	}

	private void copy(InputStream inputStream, OutputStream outputStream) throws IOException {
		byte[] buffer = new byte[32768];
		int len;

		while ((len = inputStream.read(buffer)) != -1) {
			outputStream.write(buffer, 0, len);
		}
	}

	native void keyClick(int i);
}

File: app/src/main/java/org/kbieron/iomerge/services/InputDevice.kt
package org.kbieron.iomerge.services

import android.content.Context
import android.util.Log
import java.io.*
import java.lang.Runtime.getRuntime


class InputDevice(private val context: Context) {

	private val DAEMON_NAME = "iomerge_daemon"
	private var coppied = false

	init {
		System.loadLibrary("native")
	}

	private val daemonOutFile: File = context.cacheDir.resolve(DAEMON_NAME)

	@Synchronized @Throws(IOException::class, InterruptedException::class)
	internal fun startNativeDaemon() {
		stop()

		if (!coppied) {
			FileOutputStream(daemonOutFile).use {
				copy(context.assets.open(DAEMON_NAME), it)
			}
		}

		try {
			getRuntime().exec(arrayOf("su", "-C", "chmod 777 ${daemonOutFile.absolutePath}")).waitFor()
			coppied = true
			initializePipe()

			getRuntime().exec(arrayOf("su", "-C", daemonOutFile.absolutePath)).apply {
				Thread.sleep(100)
				val exitValue = exitValue() // throws IllegalThreadStateException if process was started
				throw RuntimeException("Daemon failed to start, exited with value: $exitValue")
			}
		} catch (ignored: IllegalThreadStateException) {
			// it means process was started succesfully
		}

	}

	@Synchronized internal fun stop() {

		try {
			//noinspection SpellCheckingInspection
			getRuntime().exec(arrayOf("su", "-C", "killall $DAEMON_NAME")).waitFor()
		} catch (e: InterruptedException) {
			e.printStackTrace()
		} catch (e: IOException) {
			e.printStackTrace()
		}

	}

	internal fun emitKeyEvent(event: Int) {
		try {
			getRuntime().exec(arrayOf("su", " -C", "input keyevent $event"))
		} catch (e: IOException) {
			Log.e("IOManage", "Unable to run ", e)
		}
	}

	@Throws(IOException::class)
	private fun copy(inputStream: InputStream, outputStream: OutputStream) {
		val buffer = ByteArray(32768)
		var len: Int = inputStream.read(buffer)

		while (-1 != len) {
			outputStream.write(buffer, 0, len)
			len = inputStream.read(buffer)
		}
	}

	external private fun initializePipe()

	external fun mouseMove(x: Int, y: Int)

	external fun mousePress(button: Int)

	external fun mouseRelease(button: Int)

	external fun keyPress(c: Int)

	external fun keyRelease(c: Int)

	external fun mouseWheel(anInt: Int)

	external fun keyClick(i: Int)
}
